### PR TITLE
Added pre-commit hook and some missing types

### PR DESCRIPTION
**Pull request by muhrin**
Quality of life improvements.

## Description

* Added pre-commit hook that allows users to do `pre-commit install` and
flake8 will be run each time a commit is made
* Added 'dev' extra to e3nn.  This allows ones to do `pip install -e
.[dev]` and automatically get pytest and pre-commit (for now).  The idea
is that you automatically install libraries that everyone doing
development should have.
* Added full instructions for the pre-commit hook and running of test to
CONTRIBUTING.md
* Added a few missing type hints that I came across

## Motivation and Context
This makes automated checking during development easier and more transparent.

## How Has This Been Tested?
All tests pass.

## Screenshots
N/A

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
- [X] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [X] All new and existing tests passed.
- [x] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

**Comment by muhrin** [comment]
I've opened the PR as a work in progress to enable discussion and hear your thoughts.

@mariogeiger , do you currently run any automated code style formatting tools such as black?  If so, I could include these in the `pre-commit` hook and they would automatically be applied on each commit.  I find these kinds of things make development much smoother (not to mention easier for new contributors as they have one less thing to think about).

**Comment by mariogeiger** [comment]
I don't use automated code style formatting tools. If you find one that is compatible with our flake8 rules why not

File: e3nn/o3/_irreps.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ class Irrep(tuple):
     >>> Irrep("1o") + Irrep("2o")
     1x1o+1x2o
     """
-    def __new__(cls, l, p=None):
+    def __new__(cls, l: Union['Irrep', str, tuple], p=None):
```
Comment (mariogeiger) [inline e3nn/o3/_irreps.py:52]:
```suggestion
    def __new__(cls, l: Union[int, 'Irrep', str, tuple], p=None):
```

Comment (mariogeiger) [reply e3nn/o3/_irreps.py:52]: Tell me when you want me to review again

**Comment by muhrin** [comment]
Ok, I think _finally_ this one is ready for another round of checks :)

I think it should be all good.  I've added the flake8 linter but most of the changes I've made are cosmetic.

**Pull request merged**